generic_expression_patterns_modules/process.py

```diff
@@ -454,3 +454,256 @@ def create_all_recount2_compendium(download_dir, output_filename):
 
                 file_content = ifh.read()
                 ofh.write(file_content)
+
+
+def get_published_generic_genes(filename):
+    """Get generic genes based on input filename, which could be a URL."""
+
+    df = pd.read_csv(filename, header=0, sep="\t")
+    published_generic_genes = list(df['Gene_Name'])
+    return published_generic_genes
+
+
+def get_merged_gene_id_mapping(gene_id_filename, raw_ensembl_genes):
+    """
+    Merge genes in input gene_id file with the raw ensembl gene IDs.
+
+    Arguments:
+    - gene_id_filename: filename of input gene IDs;
+    - raw_ensembl_genes: list of strings (ensembl gene IDs)
+    """
+
+    original_gene_id_mapping = pd.read_csv(
+        gene_id_filename, header=0, sep='\t', index_col=0
+    )
+
+    # Get mapping between ensembl ids with and without version numbers.
+    # The genes in `ensembl_genes` has version numbers at the end.
+    ensembl_gene_ids = pd.DataFrame(
+        data={
+            'ensembl_version': raw_ensembl_genes,
+            'ensembl_parsed': [gene_id.split('.')[0] for gene_id in raw_ensembl_genes]
+        }
+    )
+
+    # Map ensembl gene ids with version number to gene_id_mapping
+    merged_gene_id_mapping = pd.merge(
+        original_gene_id_mapping,
+        ensembl_gene_ids,
+        left_on='ensembl_gene_id',
+        right_on='ensembl_parsed',
+        how='outer'
+    )
+
+    # Set `ensembl_version` column as the index
+    merged_gene_id_mapping.set_index('ensembl_version', inplace=True)
+
+    return merged_gene_id_mapping
+
+
+def get_renamed_columns(
```

Comment:
This function probably will give most trouble to understand. The idea is that if we read the original compendium data file is too big to load into the memory. To map the genes in it, we only need its first row (header line). After we figure out how the columns can be mapped from ensembl names to hgnc gene symbols, we will read it line by line and remove unnecessary columns.